File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/require_hook_executable.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Require hook executable rule."""

from collections.abc import Iterable

from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import is_executable, severity_for

class RequireHookExecutableRule(HarnessCheckRule):
    """Validate requireHookExecutable check."""

    category = "requireHookExecutable"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """Validate requireHookExecutable check."""
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        hooks = params.get("hooks", [])
        if not isinstance(hooks, list):
            return []
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        template = messages.get("default", "{hook} must be executable")
        return [
            Finding(
                severity_for(manifest, self.category),
                self.category,
                template.format(hook=hook),
            )
            for hook in hooks
            if isinstance(hook, str) and not is_executable(root / hook)
        ]


RULE: HarnessCheckRule = RequireHookExecutableRule()

File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/require-hook-executable.ts
#!/usr/bin/env bun
import type {
	Finding,
	HarnessCheckRule,
	HarnessManifest,
	RuleContext,
} from "../harness-check-rule";

/**
 * Require hooks to be executable.
 */
export const requireHookExecutableRule = (
	ctx: RuleContext,
): HarnessCheckRule => ({
	applies(manifest: HarnessManifest): boolean {
		const section = manifest.requireHookExecutable;
		if (typeof section !== "object" || section === null) {
			return false;
		}
		const enabled = (section as { enabled?: unknown }).enabled;
		if (enabled === false) {
			return false;
		}
		const parameters = ctx.readJsonObject(
			(section as Record<string, unknown>).parameters,
		);
		return ctx.readStringArray(parameters.hooks).length > 0;
	},

	validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
		const parameters = ctx.readJsonObject(
			ctx.readJsonObject(manifest.requireHookExecutable).parameters,
		);
		const hooks = ctx.readStringArray(parameters.hooks);
		return hooks.flatMap((hook) => {
			if (!ctx.isFile(hook)) {
				return [];
			}
			return ctx.isExecutablePath(hook)
				? []
				: [
						{
							severity: ctx.severityOf(manifest, "requireHookExecutable"),
							category: "requireHookExecutable",
							message: `${hook} must be executable`,
						},
					];
		});
	},
});
